### PR TITLE
Add Support for add endpoints of pods not still ready on kubernetes

### PR DESCRIPTION
**Pull request by amiranda-ob**
When creating a new cluster of hazelcastInstance as clients, if the creation of the instance is blocking on startup of an application on kubernetes, the app is not still ready when the discoverNodes are evaluated,, so the list of address does not contains it. If when obtaining endpoints we take into account the notReady too, it will create the instance properly.
Without this change, a pod with readynessProbe configured and blocking startup of hazelcastInstance will never start and will entry on a infinite loop checking one by one all the endpoints on the kubernetes namespace.

### CODE REVIEW

**Comment by noctarius** [comment]
Not sure I get the idea. In the end they should all find themselves using the cluster merge task. A just-yet-ready node will start as a standalone cluster and eventually will find other members that got ready and merge to a bigger cluster.

On the other hand I most probably don't want a not-yet-ready node to join a cluster.

Do I miss anything here?

**Comment by amiranda-ob** [comment]
If the application needs hazelcast to startup, for example for a spring bean injection dependency, the application will never succesfully be launched if the hazelcast cluster is not ready. This is a loop between the pod not ready and the hz cluster not ready.

In this case, the current hazelcast-kubernetes behaviour is to try connecting for each node of the namespace after not finding by label, instead of creating an standalone cluster.

In the case you mention about you probably don't want a not-yet-ready node, I think that it could be an interesting option to add a property to decide it.

In short, if I am not explained to me well, but with this case the first app of the cluster will never start because there are not available addresses, due to readynessProbe in kubernetes.

**Comment by amoAHCP** [comment]
@noctarius Sonds ok to me... but @amiranda-ob you should first check all "getNotReadyAddresses()" and then "getAddresses()"... simply to avoid "loosing" nodes... (nodes which were not ready on getAddress call and ready on getNotReadyAddresses call) the resulting list should be checked for duplicates.

**Comment by noctarius** [comment]
I think it should be up to the user to check not-ready nodes, maybe a boolean property? I see the use case when your application needs hazelcast to startup but from a cluster perspective I would, by default, the other way.

**Comment by amiranda-ob** [comment]
@amoAHCP totally agree with you, it must be more efficient by checking notReady first.
@noctarius Im agree with you too, in my case I need in startup, but probably the common usage of the library does not need to affect the startup process. A boolean property for the user to choose this option should be an interesting feature.

In any case, I think that mentioning that case in the documentation of the Readme.md could prevent very problems to the library usage. I needed to debug inside the lib to found that case that following the definition I was doing it well. Do you contemplate to add the property feature in library the roadmap?

I have been using hazelcast on an AWS cluster from several time and after migrate to kubernetes, with this minor change in a local fork, I'm very glad with the performance of hazelcast in kubernetes and the easy configuration with the service's labels. You are doing a great job!

Thank guys!

**Comment by amiranda-ob** [comment]
Ok, I'll add the changes to the PR during the day. Thanks!

**Comment by emrahkocaman** [comment]
Hi @amiranda-ob , any updates?

**Comment by googlielmo** [comment]
Hello @amiranda-ob it seems there is interest in the community about the use case addressed by this PR. See also #56 .
We would like to move forward and merge this. It would be great if you could rebase this on master and address the comments above. Thanks!

**Comment by amiranda-ob** [comment]
Hi, I've just created another PR from my fork with the configurable version of this feature.
https://github.com/hazelcast/hazelcast-kubernetes/pull/57

If this version is accepted, this PR can be closed.

Sorry for the time with this PR freezed.

**Comment by googlielmo** [comment]
Hi @amiranda-ob , thank you! So the new PR includes the code in this PR _and_ addresses the comments above, correct?

**Comment by amiranda-ob** [comment]
@googlielmo yes, thats right, the new PR includes the code on this one and adds the configuration to enable the feature as commented in this thread. So, if you want, we can close this PR.

**Comment by googlielmo** [comment]
Closing this in favour of #57
